packages/preferences/src/store/selectors.js

```diff
@@ -20,30 +20,30 @@ const withDeprecatedKeys = ( originalGet ) => ( state, scope, name ) => {
 		'showListViewByDefault',
 	];
 
+	// Check if setting is in the list and scope is either 'core/edit-post' or 'core/edit-site'
 	if (
 		settingsToMoveToCore.includes( name ) &&
 		[ 'core/edit-post', 'core/edit-site' ].includes( scope )
 	) {
-		deprecated(
-			`wp.data.select( 'core/preferences' ).get( '${ scope }', '${ name }' )`,
-			{
-				since: '6.5',
-				alternative: `wp.data.select( 'core/preferences' ).get( 'core', '${ name }' )`,
-			}
-		);
-
 		const value = originalGet( state, 'core', name );
 
-		// Hotfix for 17.5. Some of the preferences in the list above haven't been
-		// migrated to core in 17.5 (i.e: `editorMode`, https://github.com/WordPress/gutenberg/pull/57642))
-		// so we should fallback to the passed scope to avoid unexpected `undefined` values.
-		if ( value === undefined ) {
-			return originalGet( state, scope, name );
+		// If the value is found in the 'core' scope, return it and show deprecation message
+		if ( value !== undefined ) {
+			deprecated(
+				`wp.data.select( 'core/preferences' ).get( '${ scope }', '${ name }' )`,
+				{
+					since: '6.5',
+					alternative: `wp.data.select( 'core/preferences' ).get( 'core', '${ name }' )`,
+				}
+			);
+
+			return value;
 		}
 
-		return originalGet( state, 'core', name );
+		// If the value is not found in the 'core' scope, return it from the original scope without deprecation message
 	}
 
+	// Fallback to original scope for other cases
 	return originalGet( state, scope, name );
 };
 
```

Comment:
I don't really understand the change here. I was expecting just a removal of some keys from the array on line 7

Reply:
I updated the description, I'm experimenting with the following:

> Actually, a more pragmatic solution would be to skip the deprecation message if:

>  The original get call was asking for the preference in the edit-site or edit-post scope;
    and the preference cannot be found in the core scope.

> This means that the preference has not been migrated yet to core in Gutenberg, and a deprecation message should not be shown.

This basically has the same effect as if we were removing the unmigrated preferences from the list, but it does it at runtime, which makes sure the correct logic is selected for each of them just in time. I tried to get a list of non-migrated preferences, but the results were inconclusive. I used the following script that I ran from the browser console after the editor was loaded on the new post page:

```js
(function() {
    const settingsToMoveToCore = [
        'allowRightClickOverrides',
        'distractionFree',
        'editorMode',
        'fixedToolbar',
        'focusMode',
        'hiddenBlockTypes',
        'inactivePanels',
        'keepCaretInsideBlock',
        'mostUsedBlocks',
        'openPanels',
        'showBlockBreadcrumbs',
        'showIconLabels',
        'showListViewByDefault',
    ];

    const { select } = wp.data;
    const preferencesStore = select('core/preferences');

    settingsToMoveToCore.forEach(setting => {
        const valueCore = preferencesStore.get('core', setting);
        const valueEditPost = preferencesStore.get('core/edit-post', setting);
        const valueEditSite = preferencesStore.get('core/edit-site', setting);

        if (valueCore === undefined) {
            if (valueEditPost !== undefined) {
                console.log(`Setting '${setting}' found in 'core/edit-post' but not migrated to 'core' yet.`);
            } else if (valueEditSite !== undefined) {
                console.log(`Setting '${setting}' found in 'core/edit-site' but not migrated to 'core' yet.`);
            } else {
                console.log(`Setting '${setting}' is undefined in all scopes.`);
            }
        } else {
            console.log(`Setting '${setting}' is present in 'core'.`);
        }
    });
})();
```

For `release/17.5`, I got the following:

```
Setting 'allowRightClickOverrides' is present in 'core'.
Setting 'openPanels' is present in 'core'.
Setting 'showBlockBreadcrumbs' is present in 'core'.
Setting 'showIconLabels' is present in 'core'.
Setting 'showListViewByDefault' is present in 'core'.
Setting 'fixedToolbar' is present in 'core'.
Setting 'inactivePanels' is present in 'core'.

Setting 'editorMode' found in 'core/edit-post' but not migrated to 'core' yet.
Setting 'hiddenBlockTypes' found in 'core/edit-post' but not migrated to 'core' yet.

Setting 'focusMode' is undefined in all scopes.
Setting 'distractionFree' is undefined in all scopes.
Setting 'keepCaretInsideBlock' is undefined in all scopes.
Setting 'mostUsedBlocks' is undefined in all scopes.

```

At first sight, it seems I'd need to remove `editorMode` and `hiddenBlockTypes`. But then, what about `distractionFree`, `keepCareInsideBlock` and `mostUsedBlocks` and `focusMode`? I suspect they are set in other components still not executed as part of the initial editor rendering. 

Another way would be to search the code for `get` or `set` calls, but that's also error-prone.

Finally, delegating the logic to the actual proxy at runtime seems to be a safer approach, as it will be decided when the `get` is called instead of ahead of time. There might be a small performance hit compared to the other solution, but it'd be a good tradeoff, considering it's only specific to `release/17.5`. 

However, if you could clarify the cases above and help me come up with the list to remove, then I think we could just remove them from the list as you originally suggested.

What do you think?

Reply:
`undefined` value might mean `false` so for me, it shouldn't be an indication about whether we should use one scope over the other.

I don't really understand why we're complicating things:

 - If a key has been migrated in 17.5, use the proxy for it
 - If the key has not been migrated, don't use the proxy for it.

The only difference with the trunk is the list of keys, in trunk all the keys are migrated.

Reply:
What am I missing?

Reply:
> The only difference with the trunk is the list of keys, in trunk all the keys are migrated.
Right.

> if a key has been migrated in 17.5, use the proxy for it
> If the key has not been migrated, don't use the proxy for it.

What would be the best way to find out what keys have/have not been migrated in `release/17.5`? Manually by checking the list of PRs in the branch? I wanted to avoid doing it manually as it could be error-prone (e.g I don't know what PRs were opened/merged related to this migration work)

Reply:
> undefined value might mean false so for me, it shouldn't be an indication about whether we should use one scope over the other.

Thanks for clarifying. Just to document here though, `undefined` was the return value that caused the BSOD in the Site Editor, which led me to work on this hotfix: https://github.com/WordPress/gutenberg/pull/58031.

Reply:
> What would be the best way to find out what keys have/have not been migrated in release/17.5?

It looks like the non-migrated keys are `editorMode` and `hiddenBlockTypes`, see my comment in [#](https://github.com/WordPress/gutenberg/pull/58145#discussion_r1464061045). I don't know about the others that are not defined anywhere else, though.

Reply:
I'll open another PR with those two settings removed from the list. I will keep this draft, just in case.